File: uk.ac.diamond.scisoft.analysis/test/uk/ac/diamond/scisoft/analysis/dataset/SignalTest.java
/*
 * Copyright 2011 Diamond Light Source Ltd.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package uk.ac.diamond.scisoft.analysis.dataset;

import static org.junit.Assert.*;

import org.junit.Test;

public class SignalTest {

	/**
	 * Comes from the Ticket SCI#1275
	 */
	@Test
	public void testSimpleConvolveFilters() {
		AbstractDataset f = AbstractDataset.zeros(new int[] {10,10}, AbstractDataset.ARRAYFLOAT64);
		AbstractDataset g = AbstractDataset.zeros(new int[] {3,3}, AbstractDataset.ARRAYFLOAT64);
		
		
		// test against a null filter
		f.iadd(2);
		g.set(1, 1, 1);
		
		AbstractDataset result = Signal.convolve(f, g, null);
		
		assertEquals("Arrays are not equal area",  (Double) f.sum(), (Double) result.sum(), 0.1); 
		
		// set against a zeroed filter
		g.set(0, 1, 1);
		
		result = Signal.convolve(f, g, null);
		
		assertEquals("Array not zero size",  0.0, (Double) result.sum(), 0.1); 	
		
		// set against a doubling filter
		g.set(2, 1, 1);
		
		result = Signal.convolve(f, g, null);
		
		assertEquals("convolved array not twice the size",  (Double)f.sum()*2.0, (Double) result.sum(), 0.1); 
	}
	
	/**
	 * Comes from the Ticket SCI#1275
	 */
	@Test
	public void testConvolveFilters() {
		AbstractDataset f = AbstractDataset.zeros(new int[] {10,10}, AbstractDataset.ARRAYFLOAT64);
		AbstractDataset g = AbstractDataset.zeros(new int[] {3,3}, AbstractDataset.ARRAYFLOAT64);
		
		// test against a null filter
		f.iadd(1);
		g.set(1, 1, 1);
		g.set(1, 0, 1);
		g.set(1, 2, 1);
		g.set(1, 1, 0);
		g.set(1, 1, 2);
		
		AbstractDataset result = Signal.convolve(f, g, null);
		
		assertEquals("Element (0,0) is not correct",  0, result.getDouble(0,0), 0.1); 
		assertEquals("Element (1,1) is not correct",  3, result.getDouble(1,1), 0.1); 
		assertEquals("Element (2,2) is not correct",  5, result.getDouble(2,2), 0.1); 
		assertEquals("Element (3,3) is not correct",  5, result.getDouble(3,3), 0.1); 
		assertEquals("Element (6,6) is not correct",  5, result.getDouble(6,6), 0.1); 
		assertEquals("Element (9,9) is not correct",  5, result.getDouble(9,9), 0.1); 
		assertEquals("Element (10,10) is not correct",  3, result.getDouble(10,10), 0.1); 
		assertEquals("Element (11,11) is not correct",  0, result.getDouble(11,11), 0.1); 
	}

	@Test
	public void testConvolutionFilter() {
		AbstractDataset ds = DoubleDataset.arange(1000);
		AbstractDataset kernel = DoubleDataset.ones(27);
		AbstractDataset result = Signal.convolveToSameShape(ds, kernel, null);
		assertEquals(120, result.getDouble(2), 0.001);
		
		ds = ds.reshape(new int[] {10,100});
		kernel = kernel.reshape(3,9);
		result = Signal.convolveToSameShape(ds, kernel, null);
		assertEquals(ds.getDouble(5,5)*27, result.getDouble(5,5), 0.001);
		
		ds = ds.reshape(new int[] {10,10,10});
		kernel = kernel.reshape(3,3,3);
		result = Signal.convolveToSameShape(ds, kernel, null);
		assertEquals(ds.getDouble(5,5,5)*27, result.getDouble(5,5,5), 0.001);
	}

	/**
	 * Comes from the Ticket SCI#1275
	 */
	@Test
	public void testConvolve() {
		AbstractDataset one_d = AbstractDataset.zeros(new int[] {10}, AbstractDataset.ARRAYFLOAT64);
		AbstractDataset two_d = AbstractDataset.zeros(new int[] {10,10}, AbstractDataset.ARRAYFLOAT64);
		AbstractDataset three_d = AbstractDataset.zeros(new int[] {10,10,10}, AbstractDataset.ARRAYFLOAT64);
		
		@SuppressWarnings("unused")
		AbstractDataset result;
		
		try {
			result = Signal.convolve(one_d, one_d, null);
		} catch (Exception e) {
			fail("Should be able to convolve 2 1D arrays");
		}
		
		try {
			result = Signal.convolve(three_d, three_d, null);
		} catch (Exception e) {
			fail("Should be able to convolve 2 3D arrays");
		}
		
		try {
			result = Signal.convolve(one_d, two_d, null);
		} catch (IllegalArgumentException e) {
			// this is the correct exception in this case
		} catch (Exception e) {
			e.printStackTrace();
			fail("convolving 2 differnt shaped arrays should raise an IllegalArgumentException");
		}
		
		try {
			result = Signal.convolve(one_d, three_d, null);
		} catch (IllegalArgumentException e) {
			// this is the correct exception in this case
		} catch (Exception e) {
			e.printStackTrace();
			fail("convolving 2 differnt shaped arrays should raise an IllegalArgumentException");
		}
		
		try {
			result = Signal.convolve(two_d, three_d, null);
		} catch (IllegalArgumentException e) {
			// this is the correct exception in this case
		} catch (Exception e) {
			e.printStackTrace();
			fail("convolving 2 differnt shaped arrays should raise an IllegalArgumentException");
		}
		
		try {
			result = Signal.convolve(three_d, two_d, null);
		} catch (IllegalArgumentException e) {
			// this is the correct exception in this case
		} catch (Exception e) {
			e.printStackTrace();
			fail("convolving 2 differnt shaped arrays should raise an IllegalArgumentException");
		}
		
		try {
			result = Signal.convolve(three_d, one_d, null);
		} catch (IllegalArgumentException e) {
			// this is the correct exception in this case
		} catch (Exception e) {
			e.printStackTrace();
			fail("convolving 2 differnt shaped arrays should raise an IllegalArgumentException");
		}
	}

	@Test
	public void testWindows() {
		AbstractDataset w;
		w = Signal.hammingWindow(10);
		TestUtils.assertDatasetEquals(new DoubleDataset(new double[] { 0.080000, 0.187620, 0.460122, 0.770000,
				0.972259, 0.972259, 0.770000, 0.460122, 0.187620, 0.080000 }), w, 1e-5, 1e-6);

		w = Signal.hammingWindow(11);
		TestUtils.assertDatasetEquals(new DoubleDataset(new double[] { 0.08, 0.16785218, 0.39785218, 0.68214782,
				0.91214782, 1., 0.91214782, 0.68214782, 0.39785218, 0.16785218, 0.08 }), w, 1e-5, 1e-6);
	}
}

File: uk.ac.diamond.scisoft.python.test/src/scisoftpy/signal_test.py
###
# Copyright 2011 Diamond Light Source Ltd.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#   http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
###

'''
Test signal methods

Can run test in jython console (after copying jython class into user script directory):

import unittest
unittest.TestProgram(argv=["signal_test"])
'''
import unittest

import scisoftpy as np
import scisoftpy.random as rnd
import scisoftpy.signal as sig


class Test(unittest.TestCase):


    def setUp(self):
        self.da = [0.1, 1.5, 2.3, 4.1]


    def tearDown(self):
        pass

    def checkitems(self, la, ds, places=7):
        if ds.ndim == 1:
            for i in range(ds.shape[0]):
                if la == None:
                    print i, ds[i]
                else:
                    if max(la[i], ds[i]) > 1e-10:
                        self.assertAlmostEqual(la[i], ds[i], places=places)
        elif ds.ndim == 2:
            for i in range(ds.shape[0]):
                for j in range(ds.shape[1]):
                    if la == None:
                        print i,j, ds[i,j]
                    else:
                        if max(la[i][j], ds[i, j]) > 1e-10:
                            self.assertAlmostEqual(la[i][j], ds[i, j], places=places)
        elif ds.ndim == 3:
            for i in range(ds.shape[0]):
                for j in range(ds.shape[1]):
                    for k in range(ds.shape[2]):
                        if la == None:
                            print i,j,k, ds[i,j,k]
                        else:
                            if max(la[i][j][k], ds[i, j, k]) > 1e-10:
                                self.assertAlmostEqual(la[i][j][k], ds[i, j, k], places=places)

    def testCorrelate(self):
        print 'test correlate'
        da = np.array(self.da, np.float)
        ada = np.correlate(da, axes=[0])
        self.checkitems(None, ada)

        a = np.array([1, 2, 3])
        b = np.array([0, 1, 0.5])
        self.checkitems([3.5], np.correlate(a, b))
        self.checkitems([2., 3.5, 3.], np.correlate(a, b, "same"))
        self.checkitems([0.5, 2., 3.5, 3., 0.], np.correlate(a, b, "full"))
        self.checkitems([3.5], np.correlate(b, a))
        self.checkitems([2., 3.5, 3.][::-1], np.correlate(b, a, "same"))
        self.checkitems([0.5, 2., 3.5, 3., 0.][::-1], np.correlate(b, a, "full"))

    def testCorrelate2(self):
        print 'test correlate2'
        da = np.zeros([10], np.float)
        da[3] = 1
        db = da.copy()
        ada = np.correlate(da,db)
        self.checkitems(None, ada)

        db = np.zeros([10], np.float)
        db[5] = 1
        ada = np.correlate(da,db)
        self.checkitems(None, ada)

        import os
        if os.name == 'java':
            ada = sig.phasecorrelate(da,db)
            self.checkitems(None, ada)


    def testCorrelate2D(self):
        import os
        if os.name != 'java':
            return
        print 'test correlate2'
        da = np.zeros([8,8], np.float)
        da[3,3] = 20
        db = da.copy()
        ada = np.correlate(da,db)
        print ada

        da += rnd.poisson(2, size=da.shape)
        db = np.zeros([8,8], np.float)
        db[5,6] = 20
        db += rnd.poisson(2, size=db.shape)
        ada = np.correlate(da,db)
        print ada

    def testConvolve(self):
        print 'test convolve'
#        da = np.array(self.da, np.float)
#        ada = np.convolve(da, axes=[0])
#        self.checkitems(None, ada)
        a = np.array([1, 2, 3])
        b = np.array([0, 1, 0.5])

        self.checkitems([0., 1., 2.5, 4., 1.5], np.convolve(a, b))
        self.checkitems([1., 2.5, 4.], np.convolve(a, b, 'same'))
        self.checkitems([2.5], np.convolve(a, b, 'valid'))
        self.checkitems([0., 1., 2.5, 4., 1.5], np.convolve(b, a))
        self.checkitems([1., 2.5, 4.], np.convolve(b, a, 'same'))
        self.checkitems([2.5], np.convolve(b, a, 'valid'))

if __name__ == "__main__":
    #import sys
    #sys.argv = ['', 'Test.testCorrelate']
    unittest.main()
